### ISSUE TITLE
Source/exports structure

### ISSUE THREAD

**Issue by thejmazz** [WhitestormJS/whitestorm.js#211]
This issue is concerned with

- how src folders are organized
- how exports are defined for the end user

(they are related to an extent)

First off, I have never used the webpack aliases - and never will. Its too much of a leap of faith for me to blindly import things with magic names just for the sake of nicer looking imports. I'd rather have an understanding of the underlying source structure and use that, or ideally, the source is organized in a way that makes obvious sense for newcomers, or its easy to import from module directly. Currently the general structure is:

```
├── compact.js
├── components
│   ├── cameras
│   │   ├── CubeCamera.js
│   │   ├── OrthographicCamera.js
│   │   ├── PerspectiveCamera.js
│   │   └── index.js
│   ├── lights
│   │   ├── AmbientLight.js
│   │   ├── AreaLight.js
│   │   ├── DirectionalLight.js
│   │   ├── HemisphereLight.js
│   │   ├── PointLight.js
│   │   ├── SpotLight.js
│   │   └── index.js
│   └── meshes
│       ├── Box.js
│       ├── Cylinder.js
│       ├── Dodecahedron.js
│       ├── Extrude.js
│       ├── Group.js
│       ├── Icosahedron.js
│       ├── Lathe.js
│       ├── Line.js
│       ├── Model.js
│       ├── Octahedron.js
│       ├── Parametric.js
│       ├── Plane.js
│       ├── Polyhedron.js
│       ├── Ring.js
│       ├── Shape.js
│       ├── Sphere.js
│       ├── Tetrahedron.js
│       ├── Text.js
│       ├── Torus.js
│       ├── Torusknot.js
│       ├── Tube.js
│       └── index.js
├── core
│   ├── App.js
│   ├── CameraComponent.js
│   ├── Component.js
│   ├── LightComponent.js
│   ├── Loop.js
│   ├── MeshComponent.js
│   ├── ModuleManager.js
│   ├── ModuleSystem.js
│   ├── PointsComponent.js
│   ├── errors.js
│   ├── index.js
│   └── prototype
│       └── attributes.js
├── deprecation.js
├── extras
│   ├── index.js
│   ├── pass
│   │   ├── ClearPass.js
│   │   ├── MaskPass.js
│   │   ├── Pass.js
│   │   ├── RenderPass.js
│   │   ├── ShaderPass.js
│   │   ├── TexturePass.js
│   │   └── index.js
│   └── shader
│       ├── CopyShader.js
│       └── index.js
├── index.js
├── modules
│   ├── app
│   │   ├── CameraModule.js
│   │   ├── ElementModule.js
│   │   ├── EventsPatchModule.js
│   │   ├── PostProcessorModule.js
│   │   ├── RenderingModule.js
│   │   ├── ResizeModule.js
│   │   ├── SceneModule.js
│   │   ├── VirtualMouseModule.js
│   │   ├── compact.js
│   │   ├── export.js
│   │   └── index.js
│   ├── controls
│   │   ├── OrbitModule.js
│   │   ├── export.js
│   │   ├── index.js
│   │   └── lib
│   │       └── ThreeOrbitControls.js
│   ├── mesh
│   │   ├── DynamicGeometryModule.js
│   │   ├── TextureModule.js
│   │   ├── export.js
│   │   └── index.js
│   └── presets
│       ├── Preset.js
│       ├── app
│       │   └── basic.js
│       └── index.js
├── polyfill.js
└── utils
    ├── extend.js
    ├── index.js
    └── transformData.js
```
@sasha240100 what is reason for `export.js` files, if `index.js` ones are already doing the whole `export * from './thingy.js'` dance?

I don't think hiding project structure behind webpack aliases is the greatest solution. For one, that will be tricky to maintain, and it's confusing to beginners.

One approach could be to have a main index file which exports everything. Perhaps with named suffixes/prefixes, then it might look like: (similar to what currently it is like, with exceptions for *controls*, *app*).

```javascript
// This works
import { App } from 'whs'

// This does not
import { SceneModule } from 'whs'
// this does
import { app as AppModules } from 'whs'
const { SceneModule } = AppModules

// This does not
import { OrbitModule } from 'whs'
// This does
import { controls as ControlsModules } from 'whs'
const { OrbitModule } = ControlsModules

// This does
import { Box } from 'whs'
```

I think I would like it if *every single thing* could be imported directly from the module.

Just want to get this discussion started, *I think the way we export items can be improved* (consistency), hopefully we can all agree on something.

Considerations as to how this is maintained with respect to docs as well.

Furthermore, we should be careful to make sure we are not breaking spec the way we use ES6 modules. Babel follows the spec for the syntax, but the actual module resolution spec is different (and webpack 2 does not strictly follow it). And in node, you will need `.jsm` to be able to use ES6 modules (when the hit node). It would be super neat to use full ES6 in node with no transpiling for headless-gl - for the most part, it is just import/export syntax preventing the codebase running in Node 6.

###### Version:
 - [x] v2.x.x
 - [ ] v1.x.x

###### Issue type:
 - [ ] Bug
 - [x] Proposal/Enhancement
 - [ ] Question

------

<details> 
<summary> <b>Tested on: </b> </summary>

###### Desktop
   - [ ] Chrome
   - [ ] Chrome Canary
   - [ ] Chrome dev-channel
   - [ ] Firefox
   - [ ] Opera
   - [ ] Microsoft IE
   - [ ] Microsoft Edge

###### Android
   - [ ] Chrome
   - [ ] Firefox
   - [ ] Opera

###### IOS
   - [ ] Chrome
   - [ ] Firefox
   - [ ] Opera

</details>

**Comment by thejmazz** [comment]
@sasha240100 my immediate proposal is to change things like `src/modules/app/index.js` from

```javascript
import {ElementModule} from './ElementModule';
import {CameraModule} from './CameraModule';
import {RenderingModule} from './RenderingModule';
import {SceneModule} from './SceneModule';
import {ResizeModule} from './ResizeModule';
import {PostProcessorModule} from './PostProcessorModule';
import {VirtualMouseModule} from './VirtualMouseModule';
import {EventsPatchModule} from './EventsPatchModule';

export const app = {
  ElementModule,
  CameraModule,
  RenderingModule,
  SceneModule,
  ResizeModule,
  PostProcessorModule,
  VirtualMouseModule,
  EventsPatchModule
};
```

to

```javascript
export * from './ElementModule';
export * from './CameraModule';
export * from './RenderingModule';
export * from './SceneModule';
export * from './ResizeModule';
export * from './PostProcessorModule';
export * from './VirtualMouseModule';
export * from './EventsPatchModule';
```

so that they available directly from the module, for the sake of consistency (e.g. a user not knowing which "sub object" to find components from).

**Comment by sasha240100** [comment]
@thejmazz I want to keep them by sections. Any other way we can do that? (Like `AppElementModule`?)

**Comment by thejmazz** [comment]
What are the main sections?

is `App` the only thing that is going to have modules specifically for it, and any other modules will work with `Component`s?

Yeah, that's what I meant by suffixing/prefixing. So something like

- `AppElementModule`
- `ElementAppModule`

Furthermore, how's best way to name modules so that their compatibility with components is clear? Or maybe just have them throw errors if used incorrectly? Then user can check docs to see that `ElementModule` should be applied to an instance of `App`, but if used on a regular component, it throws a warning/error message. If that were the case, name would be less important - which may or maybe not is desired.

I think adding `AppModule` of some sort into name works.

**Comment by sasha240100** [comment]
@thejmazz Also consider that three.js does it for materials. **For instance:**
 - `MeshBasicMaterial`
 - `LineDashedMaterial`

Yes, for modules that **are included into whs** we can do nearly the same:
 - `AppElementModule`
 - `AppRenderingModule`

`AppXXXXXModule`

**But I'm not sure... I think that we can just drop that `WHS.app.` and leave just module name.**

**Comment by sasha240100** [comment]
Closing because this change is already included in #231
